genienlp/kfserver.py

```diff
@@ -0,0 +1,64 @@
+#
+# Copyright (c) 2018, Salesforce, Inc.
+#                     The Board of Trustees of the Leland Stanford Junior University
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+import multiprocessing
+import logging
+
+import kfserving
+
+from .util import log_model_size
+from .server import Server, init
+
+logger = logging.getLogger(__name__)
+
+
+class KFModelServer(kfserving.KFModel):
+    def __init__(self, name, args, numericalizer, model, device, confidence_estimator):
+        super().__init__(name)
+        self.server = Server(args, numericalizer, model, device, confidence_estimator)
+
+    def load(self):
+        log_model_size(logger, self.server.model, self.server.args.model)
+        self.server.model.to(self.server.device)
+        self.server.model.eval()
+        self.ready = True
+
+    def predict(self, request):
+        results = self.server.handle_request(request)
+        return {"predictions": results}
+
+
+def main(args):
+    model, device, confidence_estimator = init(args)
+    model_server = KFModelServer('nlp', args, model.numericalizer, model, device, confidence_estimator)
+    model_server.load()
+    num_workers =  min(1, multiprocessing.cpu_count() - 1)
```

Comment:
Does this work if we're in a container? I think cpu_count looks at the actual number of physical CPUs (from /proc/cpuinfo or sysconf) but we want the number of CPUs allocated to the container. I think we need a command-line argument - or hardcode this to 1. If we do GPU inference, we'll probably need this to be 1 because GPUs cannot be shared, you hit memory issues.

Reply:
I'll hardcode to 1 for now until we have some loadtest data.